source/geod24/concurrency.d

```diff
@@ -2191,296 +2107,87 @@ private
-            void lock() { while (!cas(&locked, false, true)) { Thread.yield(); } }
-            void unlock() { atomicStore!(MemoryOrder.rel)(locked, false); }
-            bool locked;
+            return msg.type != MsgType.standard;
         }
 
-        static shared SpinLock sm_lock;
-        static shared Node* sm_head;
-
-        Node* newNode(T v)
+        bool isLinkDeadMsg(ref Message msg) @safe @nogc pure nothrow
         {
-            Node* n;
-            {
-                sm_lock.lock();
-                scope (exit) sm_lock.unlock();
-
-                if (sm_head)
-                {
-                    n = cast(Node*) sm_head;
-                    sm_head = sm_head.next;
-                }
-            }
-            if (n)
-            {
-                import std.conv : emplace;
-                emplace!Node(n, v);
-            }
-            else
-            {
-                n = new Node(v);
-            }
-            return n;
+            return msg.type == MsgType.linkDead;
         }
 
-        void freeNode(Node* n)
-        {
-            // destroy val to free any owned GC memory
-            destroy(n.val);
-
-            sm_lock.lock();
-            scope (exit) sm_lock.unlock();
+        /// closed
+        bool closed;
 
-            auto sn = cast(shared(Node)*) n;
-            sn.next = sm_head;
-            sm_head = sn;
-        }
+        /// lock
+        Mutex mutex;
 
-        void put(Node* n)
-        {
-            m_count++;
-            if (!empty)
-            {
-                m_last.next = n;
-                m_last = n;
-                return;
-            }
-            m_first = n;
-            m_last = n;
-        }
+        /// size of queue
+        size_t qsize;
 
-        Node* m_first;
-        Node* m_last;
-        size_t m_count;
-    }
-}
+        /// queue of data
+        DList!Message queue;
 
-@system unittest
-{
-    import std.typecons : tuple, Tuple;
+        /// collection of send waiters
+        DList!SudoFiber sendq;
 
-    static void testfn(Tid tid)
-    {
-        receive((float val) { assert(0); }, (int val, int val2) {
-            assert(val == 42 && val2 == 86);
-        });
-        receive((Tuple!(int, int) val) { assert(val[0] == 42 && val[1] == 86); });
-        receive((Variant val) {  });
-        receive((string val) {
-            if ("the quick brown fox" != val)
-                return false;
-            return true;
-        }, (string val) { assert(false); });
-        prioritySend(tid, "done");
+        /// collection of recv waiters
+        DList!SudoFiber recvq;
     }
 
-    static void runTest(Tid tid)
-    {
-        send(tid, 42, 86);
-        send(tid, tuple(42, 86));
-        send(tid, "hello", "there");
-        send(tid, "the quick brown fox");
-        receive((string val) { assert(val == "done"); });
-    }
+    private alias StopWaitDg = void delegate ();
 
-    static void simpleTest()
+    ///
+    private struct SudoFiber
```

Comment:
Why is this called a "Sudo" fiber? It sounds like a Sudo fiber has some special permissions that no other fibers have.. Did you mean to call it a Main fiber instead?

Reply:
Can we get an answer on this one? :)